### ISSUE TITLE
Trait bound not implemented for substituted type

### ISSUE THREAD

**Issue by gregdhill** [paritytech/subxt#604]
I want to replace this struct [`VaultId<AccountId, CurrencyId: Copy>`](https://github.com/interlay/interbtc/blob/2ddb5eba6fdb42125805bf15d8a4caf30556e0ab/primitives/src/lib.rs#L66) using `substitute_type`. However, multiple compilation errors are thrown complaining that the trait bound is not satisfied.

**Comment by jsdw** [comment]
What do the errors look like? Also it might be useful to share the code for the struct you're trying to replace it with, or a small example that we can reproduce to dig into more :)

**Comment by gregdhill** [comment]
I've identified the problem, parent structs in `runtime_types` don't inherit child trait bounds.

See for example the following type generated by `subxt`:
```rust
#[derive(::subxt::ext::codec::Decode, ::subxt::ext::codec::Encode, Debug)]
pub struct Vault<_0, _1, _2, _3, _4> {
    pub id: crate::VaultId<_0, _3>,
    pub status: runtime_types::vault_registry::types::VaultStatus,
    ...
}
```

Uses the `VaultId` type declared as:
```rust
#[derive(Encode, Decode, Debug)]
pub struct VaultId<AccountId, CurrencyId: Copy> {
    ...
}
```

Which fails to compile because `Copy` is not bound to `_3`:
```
error[E0277]: the trait bound `_3: std::marker::Copy` is not satisfied
  --> src/main.rs:28:1
   |
28 | #[subxt::subxt(runtime_metadata_path = "./metadata.scale")]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `std::marker::Copy` is not implemented for `_3`
   |
note: required by a bound in `VaultId`
  --> src/main.rs:12:43
   |
12 | pub struct VaultId<AccountId, CurrencyId: Copy> {
   |                                           ^^^^ required by this bound in `VaultId`
   = note: this error originates in the derive macro `::subxt::ext::codec::Encode` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider further restricting type parameter `_3`
   |
28 | #[subxt::subxt(runtime_metadata_path = "./metadata.scale")], _3: std::marker::Copy
   |                                                            +++++++++++++++++++++++
```

**Comment by gregdhill** [comment]
The ideal solution would be to introspect trait bounds on the substituted type and propagate them upwards but I don't think this is possible.

**Comment by jsdw** [comment]
Aah I see; you substitute in a local type, but the generated type doesn't meet the type bounds of the local one.

Indeed; it's not possible to figure out at compile time what the trait bounds of some local un-annotated type are. 

A workaround for your case may be to not have the `Copy` trait defined on the struct like that and instead to define it in the relevant `impl` blocks where it's needed.

Depending on how many places `VaultId` is used, you could also override the `Vault` type itself potentially, but I can imagine that might get out of hand quite quickly!

There may be some annotation that we can add that would say "hey; if you use this type in any generated code, please propagate the following trait bounds up". I suspect it'd be quite complex though!

Let's leave this open and ponder.

**Comment by jsdw** [comment]
Just mulling over your PR and this, and I was wondering; so you have this custom type:

```rust
#[derive(Encode, Decode, Debug)]
pub struct VaultId<AccountId, CurrencyId: Copy> {
    ...
}
```

And the problem is that it's substituted in places where `CurrencyId: Copy` does not hold true, right?

So the proposed solution in the PR is to allow something like this to be substituted instead:

```
#[derive(Encode, Decode, Debug)]
pub struct VaultId {
    ...
}
```

I guess I'm wondering whether something like this would work at present:

```rust
#[derive(Encode, Decode, Debug)]
pub struct VaultId<AccountId, CurrencyId> {
    ...,
    _marker: std::marker::PhantomData<(AccountId, CurrencyId)>
}
```

Which would be equivalent to removing the generic params (less convenient though since they are rendered useless).

Or I'm wondering; why does `CurrencyId` have to implement `Copy` in the first approach at all? And if it does, how far could you get using `derive_for_type(type = "interbtc_primitives::CurrencyId", derive = "Copy, Clone")` as in your PR example to get the right trait bounds you need?

Perhaps this is a problem you've encountered with a bunch of types though, and in any case, I can see the value of the approach you took in the PR allowing substitute types to not have generic params.. it gives you more flexibility over how exactly your custom type works, which I like. I'm just curious to understand more I guess :)

**Comment by jsdw** [comment]
I forgot about this issue and opened https://github.com/paritytech/subxt/issues/666 to formulate an approach which should help to address this :)

**Comment by jsdw** [comment]
@gregdhill I wonder whether has been resolved now; #666 added some functionality around allowing generic params to be substituted in type substitutions, and then more recently I expanded that functionality so that you can pattern match on generic params and then use them nowhere, or anywhere in the replacement type (see the recent changelog update for some info on this: https://github.com/paritytech/subxt/blob/master/CHANGELOG.md).

I'll close it under the assumption it's been resolved, but lemme know if not and we can re-open :)
